pixels-turbo/pixels-worker-vhive/src/main/java/io/pixelsdb/pixels/worker/vhive/PixelsWriterStreamImpl.java

```diff
@@ -0,0 +1,830 @@
+        }
+
+        // write and flush row group content
+        int recordedRowGroupDataLen = 0;
+        try {
+            curRowGroupOffset = byteBuf.writerIndex();
+            if (curRowGroupOffset != -1) {
+                /**
+                 * Actually, this condition is always true because a ByteBuf's writerIndex is never -1.
+                 * Just keep the code for compatibility with legacy {@link PixelsWriterImpl#writeRowGroup()}, where curRowGroupOffset could be -1 when write prepare failed.
+                 */
+
+                // Issue #519: make sure to start writing the column chunks in the row group from an aligned offset.
+                int tryAlign = 0;
+                long writtenBytesBefore = writtenBytes;
+                int rowGroupDataLenPos = byteBuf.writerIndex();
+                byteBuf.writeInt(0); // write a placeholder for row group data length
+                writtenBytes += Integer.BYTES;
+                curRowGroupOffset = byteBuf.writerIndex();
+
+                while (CHUNK_ALIGNMENT != 0 && curRowGroupOffset % CHUNK_ALIGNMENT != 0 && tryAlign++ < 2) {
+                    int alignBytes = (int) (CHUNK_ALIGNMENT - curRowGroupOffset % CHUNK_ALIGNMENT);
+                    byteBuf.writeBytes(CHUNK_PADDING_BUFFER, 0, alignBytes);
+                    writtenBytes += alignBytes;
+                    curRowGroupOffset = byteBuf.writerIndex();
+                }
+                if (tryAlign > 2) {
+                    logger.warn("failed to align the start offset of the column chunks in the row group");
+                    throw new IOException("failed to align the start offset of the column chunks in the row group");
+                }
+
+                for (ColumnWriter writer : columnWriters) {
+                    byte[] columnChunkBuffer = writer.getColumnChunkContent();
+
+                    // DESIGN: Because ColumnChunkIndex does not change after the last write() or flush(),
+                    //  we have moved it from rowGroup footer (as in PixelsWriterImpl) to header here,
+                    //  which might work better with the streaming nature of this stream writer.
+                    PixelsProto.ColumnChunkIndex.Builder chunkIndexBuilder = writer.getColumnChunkIndex();
+                    chunkIndexBuilder.setChunkOffset(byteBuf.writerIndex());
+                    chunkIndexBuilder.setChunkLength(writer.getColumnChunkSize());
+                    curRowGroupIndex.addColumnChunkIndexEntries(chunkIndexBuilder.build());
+                    curRowGroupEncoding.addColumnChunkEncodings(writer.getColumnChunkEncoding().build());
+                    // DESIGN: ColumnChunkEncoding is final. Also moved it from rowGroup footer to header
+
+                    byteBuf.writeBytes(Unpooled.wrappedBuffer(columnChunkBuffer));
+                    // todo: try using byteBuf.addComponent(true, Unpooled.wrappedBuffer(columnChunkBuffer)) (and let byteBuf = Unpooled.compositeBuffer()) instead
```

Comment:
Why?

Reply:
Using `compositeBuffer.addComponent()` might attain zero-copy compared to using `writeBytes()` and improve performance. However, this also requires the added buffers (in this case, the column writer's internal buffers) to remain unchanged before they are read, which is not the case. Therefore not used for now.